crates/transaction-pool/src/pool/pending.rs

```diff
@@ -331,6 +332,148 @@ impl<T: TransactionOrdering> PendingPool<T> {
+                        for tx in offender_txs {
+                            if let Some(tx) = self.remove_transaction(tx.transaction.id()) {
+                                removed.push(tx);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // If still above threshold, reduce to limit till each offenders tx count is below
+        // max_account_slots
+        while self.len() > limit.max_txs && !offenders.is_empty() {
+            let offender = offenders.pop().unwrap();
+
+            // we do not need to decrement the offender tx_count because we're popping it and these
+            // should be removed
+            //
+            // SAFETY: all addresses from offenders are in spammers_map
+            let offender_txs =
+                spammers_map.get_mut(&offender.0).unwrap().split_off(max_account_slots);
+
+            for tx in offender_txs {
+                if let Some(tx) = self.remove_transaction(tx.transaction.id()) {
+                    removed.push(tx);
+                }
+            }
+        }
+
+        // penalize non-local txs if limit is still exceeded
+        if self.size() > limit.max_size || self.len() > limit.max_txs {
+            for tx in self.all.clone().iter() {
+                if tx.transaction.is_local() {
+                    continue
+                }
+                while self.size() > limit.max_size || self.len() > limit.max_txs {
+                    if let Some(tx) = self.pop_worst() {
+                        removed.push(tx);
+                    }
+                }
+            }
+        }
+
+        // penalize local txs at last
+        while self.size() > limit.max_size || self.len() > limit.max_txs {
+            if let Some(tx) = self.pop_worst() {
+                removed.push(tx);
+            }
+        }
+
+        removed
+    }
+
+    /// Returns account address and their txs_count that are _not local_ and whose transaction
+    /// count > `max_account_slots` limit.
+    ///
+    /// This returns a map from address to pending transactions for addresses with over
+    /// `max_account_slots` transactions.
+    pub(crate) fn get_spammers(
+        &self,
+        max_account_slots: usize,
+    ) -> HashMap<Address, Vec<PendingTransaction<T>>> {
+        let mut spammers: HashMap<Address, Vec<PendingTransaction<T>>> = HashMap::new();
+        for tx in self.all.iter() {
+            if tx.transaction.is_local() {
+                continue
+            }
+
+            // push the tx or populate with the current tx
+            let sender = tx.transaction.sender();
+            spammers
+                .entry(sender)
+                .and_modify(|txs| txs.push(tx.clone()))
+                .or_insert(vec![tx.clone()]);
+        }
+
+        // filter out accounts that are below the threshold
+        spammers.retain(|_, txs| txs.len() > max_account_slots);
```

Comment:
Because we already do not allow transactions that exceed the max account slots to enter the pool, this part of the algorithm is removed

https://github.com/paradigmxyz/reth/blob/e5362882e0da7e39912210152eebc4b08b0e6c7c/crates/transaction-pool/src/pool/txpool.rs#L1281-L1283

Now, this starts with non-local transactions, removing the highest nonce transaction for each sender, one-by-one, until either the pool is under its limits, or there are only local transactions left. Then, we do the same to local transactions.